### PR TITLE
[fix][misc] do not require encryption on system topics

### LINKED ISSUE #18897: [Bug] Not able to produce messages if transactions enabled and encryption required
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

Master, 2.11.0

### Minimal reproduce step

- Enable transactions
- Set encryption required to true in a namespace. 
- Produce a message (even not in a transaction) in that namespace

´´´
2022-12-13T00:56:26,908 - WARN  - [pulsar-io-17-4:ClientCnx@751] - [id: 0xdde12b57, L:/127.0.0.1:54275 - R:localhost/127.0.0.1:54202] Received error from server: Encryption is required in persistent://tnx/ns-prechecks/__transaction_buffer_snapshot
´´´

### What did you expect to see?

The internal topic should by pass the encryption requirement 

### What did you see instead?

Not able to produce any message since the Producer cannot be created successfuly on the client side due to that error

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by nicoloboschi**
Fixes #18897

### Motivation

System topics on user namespace may be subject to have encryption set on the producer side. However they can't be forced to be encrypted since the encryption keys are set on the client side. Also they don't contain sensitive data. 

The two main cases that I encountered are:
1. __change_events: if you enable topic level policies
2. transaction side topics

### Modifications

* Always allow messages in the system topics to not be encrypted

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

**Review by lhotari — APPROVED** [review]
LGTM

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/SystemTopic.java [hunk 1]
```diff
@@ -76,4 +76,10 @@ public boolean isCompactionEnabled() {
         // even though is not explicitly set in the policies.
         return !NamespaceService.isHeartbeatNamespace(TopicName.get(topic));
     }
+
+    @Override
+    public boolean isEncryptionRequired() {
+        // System topics are only written by the broker that can't know the encryption context.
```
Comment (michaeljmarshall) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/SystemTopic.java:82]: It's tangential, but have we explicitly stated the limitation that system topics are only written to by the broker?

**Review by michaeljmarshall — CHANGES_REQUESTED** [review]
@nicoloboschi - I support this change, but I think it needs a little more work.

This code is relevant: https://github.com/apache/pulsar/blob/5b062f37bd6688c09384abbe908f7c14e28052fc/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java#L785-L791

It is used here and in the non-partitioned topic:

https://github.com/apache/pulsar/blob/5b062f37bd6688c09384abbe908f7c14e28052fc/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L2642

If I am reading the code correctly, without updating the `Producer#checkEncryption` method, an update to a namespace policy will result in an unnecessary disconnection for producers to system topics when encryption becomes required for a namespace.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java [hunk 2]
```diff
@@ -800,7 +800,9 @@ public void checkEncryption() {
 
     public void publishTxnMessage(TxnID txnID, long producerId, long sequenceId, long highSequenceId,
                                   ByteBuf headersAndPayload, long batchSize, boolean isChunked, boolean isMarker) {
-        checkAndStartPublish(producerId, sequenceId, headersAndPayload, batchSize, null);
+        if (!checkAndStartPublish(producerId, sequenceId, headersAndPayload, batchSize, null)) {
+            return;
+        }
```
Comment (michaeljmarshall) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java:805]: Is this its own bug? It seems unrelated to encryption.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/SystemTopic.java (hunk 1 above)
Comment (nicoloboschi) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/SystemTopic.java:82]: AFAIK we haven't state that ONLY the broker (or pulsar internals) write to these topics but they for sure they do so they need this fix

**Comment by nicoloboschi** [comment]
> @nicoloboschi - I support this change, but I think it needs a little more work.
> 
> This code is relevant:
> 
> https://github.com/apache/pulsar/blob/5b062f37bd6688c09384abbe908f7c14e28052fc/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java#L785-L791
> 
> It is used here and in the non-partitioned topic:
> 
> https://github.com/apache/pulsar/blob/5b062f37bd6688c09384abbe908f7c14e28052fc/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L2642
> 
> If I am reading the code correctly, without updating the `Producer#checkEncryption` method, an update to a namespace policy will result in an unnecessary disconnection for producers to system topics when encryption becomes required for a namespace.


I've implemented isEncryptionRequired to always return false so the disconnection is never triggered. or maybe I didn't get your comments 🤔

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java (hunk 2 above)
Comment (nicoloboschi) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java:805]: yes it's a bit out of the encryption fix scope. I can open another pull request

Comment (nicoloboschi) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Producer.java]: removed. will open another pull

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 3]
```diff
@@ -1326,7 +1327,9 @@ protected void handleProducer(final CommandProducer cmdProducer) {
 
                 backlogQuotaCheckFuture.thenRun(() -> {
                     // Check whether the producer will publish encrypted messages or not
-                    if ((topic.isEncryptionRequired() || encryptionRequireOnProducer) && !isEncrypted) {
+                    if ((topic.isEncryptionRequired() || encryptionRequireOnProducer)
+                            && !isEncrypted
+                            && !SystemTopicNames.isSystemTopic(topicName)) {
```
Comment (michaeljmarshall) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:1332]: If `topic.isEncryptionRequired()` is always `false` for system topics, why do we need the check on the `topicName` on line 1332?

Comment (nicoloboschi) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:1332]: the previous condition was: topic requires encryption or producer requested encryption 
if I leave it as is, the `encryptionRequireOnProducer` will be valid also for system topics and then the `opic.isEncryptionRequired()` would be useless

Comment (michaeljmarshall) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:1332]: Oh, of course. Thanks for explaining.

**Review by michaeljmarshall — APPROVED** [review]
LGTM.

**Comment by nicoloboschi** [comment]
Since 2.11.0 system topics are enabled by defaults so I think we must include this fix in 2.11.0. Otherwise if users upgrade, they will start seeing a lot of errors in the broker

**Pull request merged**